### ISSUE TITLE
Found a bug in StableDiffusionControlNetReferencePipeline, and question about implementing StableDiffusionControlNetReferenceInpaintImg2ImgPipeline

### ISSUE THREAD

**Issue by kunc01** [huggingface/diffusers#4059]
### Model/Pipeline/Scheduler description

@okotaku Hi, I'm trying to implement StableDiffusionControlNetReferenceInpaintImg2ImgPipeline following [StableDiffusionControlNetReferencePipeline](https://github.com/huggingface/diffusers/blob/main/examples/community/stable_diffusion_controlnet_reference.py) and [StableDiffusionControlNetInpaintImg2ImgPipeline](https://github.com/huggingface/diffusers/blob/main/examples/community/stable_diffusion_controlnet_inpaint_img2img.py). My implementation and results are in [resources.zip](https://github.com/huggingface/diffusers/files/12028594/resources.zip). While [webui implementation](https://github.com/Mikubill/sd-webui-controlnet/blob/main/scripts/hook.py#L542) uses torch.zeros at the 5th input channel, it leads to blurry generated background for diffusers. Change to torch.ones leads to better result, but the reflection mainly comes from the reference image. I'm wondering how to align the result with WebUI, any help is appreciated!

### Open source status

- [ ] The model implementation is available
- [ ] The model weights are available (Only relevant if addition is not a scheduler).

**Comment by patrickvonplaten** [comment]
Let's maybe add this as a community pipeline?

**Comment by kunc01** [comment]
> 

Sure, but I'm afraid there are some bugs in my code. I haven't looked closely enough about the variations in the implementation of Unet and ControlNet between "diffusers" and "webui".
WebUI:        

    # Handle attention and AdaIn control
    for param in outer.control_params:
        if param.guidance_stopped:
            continue

        if param.used_hint_cond_latent is None:
            continue

        if param.control_model_type not in [ControlModelType.AttentionInjection]:
            continue

        ref_xt = outer.sd_ldm.q_sample(param.used_hint_cond_latent, torch.round(timesteps.float()).long())

        # Inpaint Hijack
        if x.shape[1] == 9:
            ref_xt = torch.cat([
                ref_xt,
                torch.zeros_like(ref_xt)[:, 0:1, :, :],
                param.used_hint_cond_latent
            ], dim=1)

        outer.current_style_fidelity = float(param.preprocessor['threshold_a'])
        outer.current_style_fidelity = max(0.0, min(1.0, outer.current_style_fidelity))

My diffuser implementation:

    # ref only part
    noise = randn_tensor(
        ref_image_latents.shape, generator=generator, device=device, dtype=ref_image_latents.dtype
    )
    ref_xt = self.scheduler.add_noise(
        ref_image_latents,
        noise,
        t.reshape(
            1,
        ),
    )
    ref_xt = self.scheduler.scale_model_input(ref_xt, t)

    # ! Inpaint Hijack like webUI
    # ref_xt = torch.cat([ref_xt, torch.zeros_like(ref_xt)[:, 0:1, :, :], ref_image_latents], dim=1) # terrible
    ref_xt = torch.cat([ref_xt, torch.ones_like(ref_xt)[:, 0:1, :, :], ref_image_latents], dim=1) # ok

    MODE = "write"
    self.unet(
        ref_xt,
        t,
        encoder_hidden_states=prompt_embeds,
        cross_attention_kwargs=cross_attention_kwargs,
        return_dict=False,
    )

Also, I find a bug in [StableDiffusionControlNetReferencePipeline](https://github.com/huggingface/diffusers/blob/main/examples/community/stable_diffusion_controlnet_reference.py#L326), ref_image is preprocessed into [0,1], while in [StableDiffusionReferencePipeline](https://github.com/huggingface/diffusers/blob/main/examples/community/stable_diffusion_reference.py#L320) and webui ref_image is preprocessed into [-1,1].
